### PR TITLE
update document for validation.

### PR DESCRIPTION
**Pull request by vwxyzh**
@chenkennt @superyyrrzz @hellosnow @ansyral @qinezh

### CODE REVIEW

**Comment by chenkennt** [comment]
> [Export(typeof(IInputMetadataValidator))]

update this part? now you need to give it a name and enable it in md.style

---
Refers to: Documentation/tutorial/validate_your_markdown_files.md:228 in 34aeb2c. [](commit_id = 34aeb2c92dfc889f8e22b27ebc6dcda70375174a, deletion_comment = False)

File: Documentation/tutorial/validate_your_markdown_files.md [hunk 1]
```diff
@@ -235,3 +239,102 @@ public class MyInputMetadataValidator : IInputMetadataValidator
...
+
+    ```json
+    {
+      "metadataRules": [
+        { "contractName": "hello_is_not_valid", "disable": false }
+      ]
+    }
+    ```
+
+## Advance: Share your rules
```
Comment (chenkennt) [inline Documentation/tutorial/validate_your_markdown_files.md]: >Advance [](start = 3, length = 7)

Advanced

File: Documentation/tutorial/validate_your_markdown_files.md [hunk 2]
```diff
@@ -235,3 +239,102 @@ public class MyInputMetadataValidator : IInputMetadataValidator
...
+
+Some users have a lot of document projects, and want to share validations for all of them, and don't want to write `md.style` file repeatly.
```
Comment (chenkennt) [inline Documentation/tutorial/validate_your_markdown_files.md]: >repeatly [](start = 131, length = 8)

repeatedly

File: Documentation/tutorial/validate_your_markdown_files.md [hunk 3]
```diff
@@ -38,20 +39,23 @@ Then when anyone write `<H1>` or `<H2>` in Markdown file, it will give a warning
 You can use the following proprties to configure the HTML tag rule:
 
 1.  `tagNames` is the list of HTML tag names to validate, *required*, *case-insensitive*.
-2.  `behavior` defines the behavior when the HTML tag is met, *required*. Its value can be following:
+2.  `relation` is a choice for `tagNames`:
```
Comment (hellosnow) [inline Documentation/tutorial/validate_your_markdown_files.md]: >is a choice [](start = 15, length = 11)

is a choice -> is optional?

**Comment by vwxyzh** [comment]
> [Export(typeof(IInputMetadataValidator))]

it cannot be disable.

---
In reply to: [238470658](https://github.com/dotnet/docfx/pull/587#issuecomment-238470658) [](ancestors = 238470658)

---
Refers to: Documentation/tutorial/validate_your_markdown_files.md:228 in 34aeb2c. [](commit_id = 34aeb2c92dfc889f8e22b27ebc6dcda70375174a, deletion_comment = False)

File: Documentation/tutorial/validate_your_markdown_files.md [hunk 4]
```diff
@@ -235,3 +239,106 @@ public class MyInputMetadataValidator : IInputMetadataValidator
...
+}
+```
+
+Then `test` is the category name (from file name) for three rules, and apply different `id` for each rule, they are `heading`, `rules` and `hello`.
+
+When you build document with this template, all rules will be active when `disable` property is `false`.
+
+### Config rules
+
+Some rules need to be enable/disable in some special document project.
```
Comment (superyyrrzz) [inline Documentation/tutorial/validate_your_markdown_files.md]: enabled/disabled

**Comment by superyyrrzz** [comment]
:+1:

**Comment by hellosnow** [comment]
> 1.  Create a project in your code editor (e.g. visual studio).

minor: Visual Studio

---
Refers to: Documentation/tutorial/validate_your_markdown_files.md:72 in d328d89. [](commit_id = d328d89bfd7ebe3e647e10ce7a250f79bf3bdadd, deletion_comment = False)

**Comment by hellosnow** [comment]
> The different between `ReportError` and throw `DocumentException` is throw exception will stop the build immediately but `ReportError` won't stop build but will eventually fail the build after rules are run.

throwing?

---
Refers to: Documentation/tutorial/validate_your_markdown_files.md:186 in d328d89. [](commit_id = d328d89bfd7ebe3e647e10ce7a250f79bf3bdadd, deletion_comment = False)

**Comment by hellosnow** [comment]
> And we allow add some plug-ins to validate metadata written in markdown files.

allow to add

---
Refers to: Documentation/tutorial/validate_your_markdown_files.md:208 in d328d89. [](commit_id = d328d89bfd7ebe3e647e10ce7a250f79bf3bdadd, deletion_comment = False)

**Comment by hellosnow** [comment]
> Metadata is coming multiple sources, the following metadata will be validated during build: 

coming from?

---
Refers to: Documentation/tutorial/validate_your_markdown_files.md:212 in d328d89. [](commit_id = d328d89bfd7ebe3e647e10ce7a250f79bf3bdadd, deletion_comment = False)

File: Documentation/tutorial/validate_your_markdown_files.md [hunk 5]
```diff
@@ -235,3 +239,106 @@ public class MyInputMetadataValidator : IInputMetadataValidator
 ```
 
 Enable metadata rule is same as other rules, just copy the assemblies to the `plugins` of your template folder and run `docfx`.
+
+### Create configurable metadata validation plug-ins
+
+There are two steps to create a metadata validation:
```
Comment (hellosnow) [inline Documentation/tutorial/validate_your_markdown_files.md]: >create a metadata validation [](start = 23, length = 28)

create a metadata validation -> validator?

File: Documentation/tutorial/validate_your_markdown_files.md [hunk 6]
```diff
@@ -235,3 +239,106 @@ public class MyInputMetadataValidator : IInputMetadataValidator
...
+   "rules": {
+      "code": "code_snippet_should_be_csharp"
+   },
+   "metadataRules": {
+     "hello": { "contractName": "hello_is_not_valid", "disable": true }
+   }
+}
+```
+
+Then `test` is the category name (from file name) for three rules, and apply different `id` for each rule, they are `heading`, `rules` and `hello`.
```
Comment (hellosnow) [inline Documentation/tutorial/validate_your_markdown_files.md]: have a question, the id is 'rules' or 'code'?

Comment (vwxyzh) [inline Documentation/tutorial/validate_your_markdown_files.md]: mistake.

---
In reply to: [74188550](https://github.com/dotnet/docfx/pull/587#discussion_r74188550) [](ancestors = 74188550)

**Comment by chenkennt** [comment]
> [Export(typeof(IInputMetadataValidator))]

then it will enabled by default? seems to be different from our design: only rules with "default" name or listed in md.style will be enabled.

---
In reply to: [238746737](https://github.com/dotnet/docfx/pull/587#issuecomment-238746737) [](ancestors = 238746737,238470658)

---
Refers to: Documentation/tutorial/validate_your_markdown_files.md:228 in 34aeb2c. [](commit_id = 34aeb2c92dfc889f8e22b27ebc6dcda70375174a, deletion_comment = False)

**Comment by hellosnow** [comment]
:shipit:

**Comment by vwxyzh** [comment]
> [Export(typeof(IInputMetadataValidator))]

I can remove it, but it will be breaking change.

---
In reply to: [238799318](https://github.com/dotnet/docfx/pull/587#issuecomment-238799318) [](ancestors = 238799318,238746737,238470658)

---
Refers to: Documentation/tutorial/validate_your_markdown_files.md:228 in 34aeb2c. [](commit_id = 34aeb2c92dfc889f8e22b27ebc6dcda70375174a, deletion_comment = False)

**Comment by vwxyzh** [comment]
> [Export(typeof(IInputMetadataValidator))]

And they are totally different things, tag/token validation only work when markdown engine is dfm, metadata validation for any markdown engine.

---
In reply to: [239058321](https://github.com/dotnet/docfx/pull/587#issuecomment-239058321) [](ancestors = 239058321,238799318,238746737,238470658)

---
Refers to: Documentation/tutorial/validate_your_markdown_files.md:228 in 34aeb2c. [](commit_id = 34aeb2c92dfc889f8e22b27ebc6dcda70375174a, deletion_comment = False)

**Comment by qinezh** [comment]
:+1:

**Pull request merged**